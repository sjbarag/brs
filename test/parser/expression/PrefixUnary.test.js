const brs = require("brs");
const { Lexeme } = brs.lexer;
const { Int32, BrsBoolean } = brs.types;

const { token, identifier, EOF } = require("../ParserTests");

describe("parser prefix unary expressions", () => {
    let parser;

    beforeEach(() => {
        parser = new brs.parser.Parser();
    });

    it("parses unary 'not'", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.Equal, "="),
            token(Lexeme.Not, "not"),
            token(Lexeme.True, "true", BrsBoolean.True),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("parses consecutive unary 'not'", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.Equal, "="),
            token(Lexeme.Not, "not"),
            token(Lexeme.Not, "not"),
            token(Lexeme.Not, "not"),
            token(Lexeme.Not, "not"),
            token(Lexeme.Not, "not"),
            token(Lexeme.True, "true", BrsBoolean.True),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("parses unary '-'", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.Equal, "="),
            token(Lexeme.Minus, "-"),
            token(Lexeme.Integer, "5", new Int32(5)),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("parses consecutive unary '-'", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.Equal, "="),
            token(Lexeme.Minus, "-"),
            token(Lexeme.Minus, "-"),
            token(Lexeme.Minus, "-"),
            token(Lexeme.Minus, "-"),
            token(Lexeme.Minus, "-"),
            token(Lexeme.Integer, "5", new Int32(5)),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("parses unary '+'", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.Equal, "="),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Integer, "5", new Int32(5)),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("parses consecutive unary '+'", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.Equal, "="),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Integer, "5", new Int32(5)),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("parses consecutive mixed unary operators", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.Equal, "="),
            token(Lexeme.Plus, "-"),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Plus, "-"),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Plus, "-"),
            token(Lexeme.Plus, "+"),
            token(Lexeme.Integer, "5", new Int32(5)),
            EOF,
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    test("location tracking", () => {
        /**
         *    0   0   0   1   1   2
         *    0   4   8   2   6   0
         *  +----------------------
         * 1| _false = not true
         */
        let { statements, errors } = parser.parse([
            {
                kind: Lexeme.Identifier,
                text: "_false",
                isReserved: false,
                location: {
                    start: { line: 1, column: 0 },
                    end: { line: 1, column: 6 },
                },
            },
            {
                kind: Lexeme.Equal,
                text: "=",
                isReserved: false,
                location: {
                    start: { line: 1, column: 7 },
                    end: { line: 1, column: 8 },
                },
            },
            {
                kind: Lexeme.Not,
                text: "not",
                isReserved: true,
                location: {
                    start: { line: 1, column: 9 },
                    end: { line: 1, column: 12 },
                },
            },
            {
                kind: Lexeme.True,
                text: "true",
                literal: BrsBoolean.True,
                isReserved: true,
                location: {
                    start: { line: 1, column: 13 },
                    end: { line: 1, column: 17 },
                },
            },
            {
                kind: Lexeme.Eof,
                text: "\0",
                isReserved: false,
                location: {
                    start: { line: 1, column: 17 },
                    end: { line: 1, column: 18 },
                },
            },
        ]);

        expect(errors).toEqual([]);
        expect(statements.length).toBe(1);
        expect(statements[0].value.location).toEqual({
            start: { line: 1, column: 9 },
            end: { line: 1, column: 17 },
        });
    });
});
