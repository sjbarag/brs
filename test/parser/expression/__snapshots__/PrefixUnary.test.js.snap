// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser prefix unary expressions parses consecutive unary '-' 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "=",
      },
    },
    "type": "Assignment",
    "value": Unary {
      "operator": Object {
        "isReserved": false,
        "kind": "Minus",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "-",
      },
      "right": Unary {
        "operator": Object {
          "isReserved": false,
          "kind": "Minus",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "-",
        },
        "right": Unary {
          "operator": Object {
            "isReserved": false,
            "kind": "Minus",
            "literal": undefined,
            "location": Object {
              "end": Object {
                "column": -9,
                "line": -9,
              },
              "start": Object {
                "column": -9,
                "line": -9,
              },
            },
            "text": "-",
          },
          "right": Unary {
            "operator": Object {
              "isReserved": false,
              "kind": "Minus",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": -9,
                  "line": -9,
                },
                "start": Object {
                  "column": -9,
                  "line": -9,
                },
              },
              "text": "-",
            },
            "right": Unary {
              "operator": Object {
                "isReserved": false,
                "kind": "Minus",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "text": "-",
              },
              "right": Literal {
                "_location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "type": "Literal",
                "value": Int32 {
                  "kind": 4,
                  "value": 5,
                },
              },
              "type": "Unary",
            },
            "type": "Unary",
          },
          "type": "Unary",
        },
        "type": "Unary",
      },
      "type": "Unary",
    },
  },
]
`;

exports[`parser prefix unary expressions parses consecutive unary 'not' 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "=",
      },
    },
    "type": "Assignment",
    "value": Unary {
      "operator": Object {
        "isReserved": true,
        "kind": "Not",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "not",
      },
      "right": Unary {
        "operator": Object {
          "isReserved": true,
          "kind": "Not",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "not",
        },
        "right": Unary {
          "operator": Object {
            "isReserved": true,
            "kind": "Not",
            "literal": undefined,
            "location": Object {
              "end": Object {
                "column": -9,
                "line": -9,
              },
              "start": Object {
                "column": -9,
                "line": -9,
              },
            },
            "text": "not",
          },
          "right": Unary {
            "operator": Object {
              "isReserved": true,
              "kind": "Not",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": -9,
                  "line": -9,
                },
                "start": Object {
                  "column": -9,
                  "line": -9,
                },
              },
              "text": "not",
            },
            "right": Unary {
              "operator": Object {
                "isReserved": true,
                "kind": "Not",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "text": "not",
              },
              "right": Literal {
                "_location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "type": "Literal",
                "value": BrsBoolean {
                  "kind": 2,
                  "value": true,
                },
              },
              "type": "Unary",
            },
            "type": "Unary",
          },
          "type": "Unary",
        },
        "type": "Unary",
      },
      "type": "Unary",
    },
  },
]
`;

exports[`parser prefix unary expressions parses unary '-' 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "=",
      },
    },
    "type": "Assignment",
    "value": Unary {
      "operator": Object {
        "isReserved": false,
        "kind": "Minus",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "-",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 5,
        },
      },
      "type": "Unary",
    },
  },
]
`;

exports[`parser prefix unary expressions parses unary 'not' 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "=",
      },
    },
    "type": "Assignment",
    "value": Unary {
      "operator": Object {
        "isReserved": true,
        "kind": "Not",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "not",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": BrsBoolean {
          "kind": 2,
          "value": true,
        },
      },
      "type": "Unary",
    },
  },
]
`;
